### PR TITLE
[RFC] cmake: add ccache support

### PR DESCRIPTION
**Pull request by Andarwinux**
For faster build, LTO can't enabled at the moment, so ccache is used instead. This is a draft, ported from an experimental [branch ](https://github.com/zhongflyTeam/mpv-winbuild/tree/sccache)of mpv-winbuild, it allow local build and CI build to use ccache.

Not sure if putting the ccache dir in MINGW_INSTALL_PREFIX is appropriate, and if the 1GB limit is reasonable. Also, not sure if ccache can cause build errors for various reasons, which need to be evaluated.

### CODE REVIEW

**Comment by zhongfly** [comment]
I still don't know if ccache is appropriate, this is just my personal guess.
And I would suggest adding a judgment like mallloc that is only really used when ccache is already installed, and keeping the ENABLE_CCACHE option

**Comment by zhongfly** [comment]
By the way, in my experience, enabling LTO for llvm does not speed up builds, almost no change.It's probably false that toolchain LTO will result in significant improvements.

**Comment by Andarwinux** [comment]
> By the way, in my experience, enabling LTO for llvm does not speed up builds, almost no change.It's probably false that toolchain LTO will result in significant improvements.

LTO for LLVM and LTO for the toolchain are different things, one enable LTO for clang and lld, the other for mingw and libc++. but it's true that LTO for LLVM itself is not much of an improvement. [PGO](https://github.com/mstorsjo/llvm-mingw/issues/383) is the future.

**Comment by zhongfly** [comment]
Don't hardcode ccache options in exec.in.
All these options should be set by the user.

**Comment by shinchiro** [comment]
Not interested as this might give some build errors

**Comment by Andarwinux** [comment]
I haven't found any ccache-induced bugs so far, but it does reduce build time significantly.

File: exec.in [hunk 1]
```diff
@@ -5,5 +5,8 @@ export PKG_CONFIG_LIBDIR="@MINGW_INSTALL_PREFIX@/lib/pkgconfig"
 export RUSTUP_HOME="@RUSTUP_LOCATION@/.rustup"
 export CARGO_HOME="@RUSTUP_LOCATION@/.cargo"
 export LD_PRELOAD="@MALLOC@"
+export CCACHE_DIR="@MINGW_INSTALL_PREFIX@/ccache"
+export CCACHE_MAXSIZE="1G"
+export CCACHE_DIRECT="true"
```
Comment (shinchiro) [inline exec.in]: This var only used by clang. Move the vars to llvm-compiler.in etch

Comment (Andarwinux) [reply exec.in]: I also plan to migrate to the wrapper for gcc in order to add ccache and gc flags support for gcc. It should be done in a few days.

**Comment by Andarwinux** [comment]
On my 16-core machine, it takes about a minute and a half to do the second clean build using gcc.

File: toolchain/llvm/llvm-compiler.in [hunk 2]
```diff
@@ -17,7 +17,11 @@ fi
 
 if [ "$CONF" == "1" ]; then
     SKIP_OPT="-O0 -fno-lto"
+else
+    if [ "@ENABLE_CCACHE@" = "ON" ]; then
+        CCACHE="ccache @CCACHE_OPTIONS@"
```
Comment (shinchiro) [inline toolchain/llvm/llvm-compiler.in]:
```suggestion
        CCACHE_ENV='CCACHE_DIR="@MINGW_INSTALL_PREFIX@/ccache" CCACHE_MAXSIZE="500M"  CACHE_DIRECT="true"'
        CCACHE="$CCACHE_ENV ccache @CCACHE_OPTIONS@"
```
I would prefer this way instead putting in exec

**Comment by shinchiro** [comment]
Split gcc wrapper commit from this

File: toolchain/llvm/llvm-compiler.in (hunk 2 above)
Comment (Andarwinux) [reply toolchain/llvm/llvm-compiler.in]: Unfortunately, it doesn't work. ccache treats these as compiler args.

Comment (shinchiro) [reply toolchain/llvm/llvm-compiler.in]: Based on [this](https://github.com/ccache/ccache/blob/master/doc/MANUAL.adoc#configuration), you dont need to export the variables. So this is possible:

`ccache cache_dir="@MINGW_INSTALL_PREFIX@/ccache" max_size=500M gcc -c example.c`

Default value for `direct_mode ` is true, so dont need to pass it

Comment (Andarwinux) [reply toolchain/llvm/llvm-compiler.in]: I don't know why, but cache_dir doesn't seem to work, ccache still saves the cache in ~/.cache/ccache

**Pull request merged**